src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,375 @@
+createVersionHistoryTable c =
+  exec_
+    c
+    "CREATE TABLE VersionHistory (version UNSIGNED BIGINT PRIMARY KEY,blockheight UNSIGNED BIGINT);"
+
+createVersionedTablesTable  :: Database -> IO ()
+createVersionedTablesTable c =
+  exec_
+    c
+    "CREATE TABLE VersionedTables (tableid TEXT PRIMARY KEY\
+  \ , blockheight UNSIGNED BIGINT\
+  \ , version UNSIGNED BIGINT);"
+
+{-
+
+for ints, investigate which int?
+
+  history of BH pairs  -- BlockHistory
+
+  schema BlockHistory
+
+  create table BlockHistory
+    (block  INT PRIMARY KEY
+    ,**header    varchar<256>) ** investigate best type for hashes
+    (BLOB seems to be the answer.)
+
+
+  history of BV pairs -- VersionHistory
+
+  schema VersionHistory
+
+  create table VersionHistory
+    (version  INT PRIMARY KEY
+    ,block    INT)
+
+  table management table -- VersionedTables
+
+  schema for VersionedTables
+
+  create table VersionedTables
+        (tableid   varchar<256> PRIMARY KEY
+        , block    INT
+        ,version   INT
+        , touchedBlock INT ** maybe
+        , touchedVersion INT ** maybe
+        )
+
+  tableid (primary key)
+
+  touched table
+
+
+
+-}
+
+detectVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockHash
+  -> IO (Either String OnVersionChange)
+detectVersionChange c bRestore hsh  = do
+  let sqlBtoBH [[SInt bh]] = BlockHeight (fromIntegral bh)
+      sqlBtoBH _ = error "sqlBtoBH: wat"
+      sqlVtoV [[SInt version]] = ReorgVersion version
+      sqlVtoV _ = error "sqlVtoV: wat"
+  bCurrent <- sqlBtoBH <$>
+    qry_ c "SELECT max(blockheight) AS current_block_height FROM BlockHistory;" [RInt]
+  vCurrent <- sqlVtoV <$>
+    qry_ c "SELECT max(version) AS current_version FROM VersionHistory;" [RInt]
+
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <-
+    qry c "SELECT COUNT(*)\
+           \ FROM BlockHistory\
+           \ WHERE blockheight = (?)\
+           \ AND hash = (?);" [SInt $ fromIntegral $ pred bRestore, SBlob (encode hsh)] [RInt]
+
+  if historyInvariant /= [[SInt 1]]
+    then return $ Left "HistoryInvariantViolation"
+  -- enforce invariant that B_restore is not greater than B_current + 1
+    else return $ case compare bRestore (bCurrent + 1) of
+        GT -> Left "RestoreInvariantViolation"
+        EQ -> Right $ NormalOperation (BlockVersion bRestore vCurrent)
+        LT -> Right $ Forking (BlockVersion bRestore (vCurrent + 1))
+
+data OnVersionChange
+  = Forking !BlockVersion
+  | NormalOperation !BlockVersion
+
+onVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockVersion
+  -> IO (Either String (BlockVersion, SnapshotName))
+onVersionChange c bRestore (BlockVersion bCurrent vCurrent) =
+  case compare bRestore (bCurrent + 1) of
+
+    -- Probably better to throw an exception here.
+    GT -> return $ Left "bRestore should not be greater than bCurrent + 1"
+
+    -- normal operation
+    EQ -> do
+      name <- mkSnapshot c
```

Comment:
You'll probably want to start this elsewhere, plus you'll want to run the version maintenance case in it's own transaction (outside of a tx, sqlite will autocommit on each stmt which is slow).